File: src/main/java/io/github/flegacy/flms/items/ItemLibrary.java
package io.github.flegacy.flms.items;

import io.github.flegacy.flms.FLMS;
import io.github.flegacy.flms.utils.TextConstants;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeModifier;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.PotionMeta;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class ItemLibrary {

	private static ItemLibrary instance;
	private final FLMSEnchanter enchanter;

    private static final String FLMS_WAND_TAG = "flms_wand";

    private final ItemStack wand;
    private final ItemStack emptyGlass;
    private final ItemStack leftArrow;
    private final ItemStack rightArrow;
    private final ItemStack backButton;
    private final ItemStack toolConfigIcon;
    private final ItemStack blockConfigIcon;
    private final ItemStack regionConfigIcon;
	private final ItemStack effectConfigIcon;

    protected final NamespacedKey flmsItemKey;
	protected final NamespacedKey enchantKey;

    public ItemLibrary(FLMS plugin) {
		if (instance != null)
			throw new IllegalStateException("There can only be one ItemLibrary instance.");
		instance = this;
		enchanter = new FLMSEnchanter(this);

        flmsItemKey = new NamespacedKey(plugin, "flms_item");
		enchantKey = new NamespacedKey(plugin, "flms_efficiency");

        wand = new ItemStack(Material.GOLDEN_AXE);
        final ItemMeta wandMeta = wand.getItemMeta();
        wandMeta.customName(TextConstants.miniMessage(TextConstants.FLMS_ORANGE + "<!i>FLMS Wand"));
        wandMeta.lore(TextConstants.messageList(TextConstants.FLMS_YELLOW + "<!i>Hold and right-click to use!"));
        wandMeta.setEnchantmentGlintOverride(true);
        final AttributeModifier attackSpeedModifier = new AttributeModifier(flmsItemKey, 999,
                AttributeModifier.Operation.ADD_NUMBER);
        wandMeta.addAttributeModifier(Attribute.ATTACK_SPEED, attackSpeedModifier);
        wandMeta.setUnbreakable(true);
        wandMeta.addItemFlags(ItemFlag.HIDE_ATTRIBUTES, ItemFlag.HIDE_UNBREAKABLE, ItemFlag.HIDE_DESTROYS);
        wandMeta.getPersistentDataContainer().set(flmsItemKey, PersistentDataType.STRING, FLMS_WAND_TAG);
        wand.setItemMeta(wandMeta);

        emptyGlass = new ItemStackBuilder(Material.GRAY_STAINED_GLASS_PANE)
                .setDisplayName("")
                .addItemFlags(ItemFlag.values())
                .build();

        leftArrow = new ItemStackBuilder(Material.ARROW)
                .setDisplayName(TextConstants.FLMS_ORANGE + "<- Previous Page")
                .setLore(TextConstants.FLMS_YELLOW + "<b>CLICK TO VIEW")
                .build();

        rightArrow = new ItemStackBuilder(Material.ARROW)
                .setDisplayName(TextConstants.FLMS_ORANGE + "Next Page ->")
                .setLore(TextConstants.FLMS_YELLOW + "<b>CLICK TO VIEW")
                .build();

        backButton = new ItemStackBuilder(Material.BARRIER)
                .setDisplayName(TextConstants.FLMS_RED + "<- Go Back")
                .setLore(TextConstants.FLMS_YELLOW + "<b>CLICK TO VIEW")
                .build();

        blockConfigIcon = new ItemStackBuilder(Material.BEDROCK)
                .setDisplayName(TextConstants.FLMS_ORANGE + "Edit Custom Blocks")
                .setLore(
                        "",
                        TextConstants.FLMS_LIGHT_YELLOW + "Manipulate all types of blocks by",
                        TextConstants.FLMS_LIGHT_YELLOW + "setting their hardness values, changing",
                        TextConstants.FLMS_LIGHT_YELLOW + "what tools are the best at breaking them,",
                        TextConstants.FLMS_LIGHT_YELLOW + "and controlling what items they drop.",
                        "",
                        TextConstants.FLMS_YELLOW + "<b>CLICK TO OPEN")
                .build();
        
        toolConfigIcon = new ItemStackBuilder(Material.GOLDEN_PICKAXE)
                .setDisplayName(TextConstants.FLMS_ORANGE + "Edit Custom Tools")
                .addItemFlags(ItemFlag.HIDE_ATTRIBUTES)
                .setLore(
                        "",
                        TextConstants.FLMS_LIGHT_YELLOW + "Create your own tools from any",
                        TextConstants.FLMS_LIGHT_YELLOW + "item in the game and define how",
                        TextConstants.FLMS_LIGHT_YELLOW + "strong they are with breaking power",
                        TextConstants.FLMS_LIGHT_YELLOW + "and enchantments.",
                        "",
                        TextConstants.FLMS_YELLOW + "<b>CLICK TO OPEN")
                .build();

        regionConfigIcon = new ItemStackBuilder(Material.GLOBE_BANNER_PATTERN)
                .setDisplayName(TextConstants.FLMS_ORANGE + "Edit Regions")
                .setLore(
                        "",
                        TextConstants.FLMS_LIGHT_YELLOW + "If you only want certain parts of",
                        TextConstants.FLMS_LIGHT_YELLOW + "the world to be breakable by players,",
                        TextConstants.FLMS_LIGHT_YELLOW + "you can manage that here.",
                        "",
                        TextConstants.FLMS_YELLOW + "<b>CLICK TO OPEN")
		        .addItemFlags(ItemFlag.HIDE_ATTRIBUTES)
                .build();

	    effectConfigIcon = new ItemStackBuilder(Material.POTION)
			    .setDisplayName(TextConstants.FLMS_ORANGE + "Edit Player Effects")
			    .addItemFlags(ItemFlag.values())
			    .setLore(
					    "",
					    TextConstants.FLMS_LIGHT_YELLOW + "Vanilla haste and mining fatigue",
					    TextConstants.FLMS_LIGHT_YELLOW + "won't work with this plugin, so you",
					    TextConstants.FLMS_LIGHT_YELLOW + "can modify those here instead or with",
					    TextConstants.FLMS_LIGHT_YELLOW + "the /flms effect command",
					    "",
					    TextConstants.FLMS_YELLOW + "<b>CLICK TO OPEN")
			    .build();
	    PotionMeta potionMeta = (PotionMeta) effectConfigIcon.getItemMeta();

    }

    public ItemStack getWand() {
        return wand.clone();
    }

    public boolean isWand(@NotNull ItemStack item) {
        if (!item.hasItemMeta())
            return false;
        final PersistentDataContainer container = item.getItemMeta().getPersistentDataContainer();
        if (!container.has(flmsItemKey))
            return false;
        return Objects.equals(container.get(flmsItemKey, PersistentDataType.STRING), FLMS_WAND_TAG);
    }

    public ItemStack getEmptyGlass() {
        return emptyGlass.clone();
    }

    public ItemStack getLeftArrow() {
        return leftArrow.clone();
    }

    public ItemStack getRightArrow() {
        return rightArrow.clone();
    }

    public ItemStack getBackButton() {
        return backButton.clone();
    }

	public ItemStack getToolConfigIcon() {
		return toolConfigIcon.clone();
	}

	public ItemStack getBlockConfigIcon() {
		return blockConfigIcon.clone();
	}

	public ItemStack getRegionConfigIcon() {
		return regionConfigIcon.clone();
	}

	public ItemStack getEffectConfigIcon() {
		return effectConfigIcon.clone();
	}

	public FLMSEnchanter getEnchanter() {
		return enchanter;
	}
}

File: src/main/kotlin/io/github/flegacy/flms/items/ItemLibrary.kt
package io.github.flegacy.flms.items

import io.github.flegacy.flms.FLMS
import io.github.flegacy.flms.util.FLMS_GRAY
import io.github.flegacy.flms.util.FLMS_GREEN
import io.github.flegacy.flms.util.FLMS_LIGHT_GREEN
import io.github.flegacy.flms.util.FLMS_LIGHT_RED
import io.github.flegacy.flms.util.FLMS_LIGHT_YELLOW
import io.github.flegacy.flms.util.FLMS_ORANGE
import io.github.flegacy.flms.util.FLMS_RED
import io.github.flegacy.flms.util.FLMS_WHITE
import io.github.flegacy.flms.util.FLMS_YELLOW
import io.github.flegacy.flms.util.msgFormat
import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.persistence.PersistentDataType

private const val FLMS_WAND_TAG = "flms_wand"
private const val EFFICIENCY_PREFIX = "⛏ Efficiency "

class ItemLibrary(plugin: FLMS) {

    private val itemKey = NamespacedKey(plugin, "flms_item")
    private val effKey = NamespacedKey(plugin, "flms_efficiency")
    val toolKey = NamespacedKey(plugin, "flms_tool")

    val enchanter = Enchanter()

    fun isWand(item: ItemStack): Boolean {
        if (item.type.isAir)
            return false
        if (!item.itemMeta.persistentDataContainer.has(itemKey))
            return false
        return item.itemMeta.persistentDataContainer.get(itemKey, PersistentDataType.STRING)
            .equals(FLMS_WAND_TAG)
    }

    fun wand(): ItemStack {
        val wand = ItemStackBuilder(Material.GOLDEN_AXE)
            .name("${FLMS_ORANGE}FLMS Wand")
            .lore("${FLMS_YELLOW}Hold and right-click to use!")
            .unbreakable(true)
            .glow(true)
            .flag(*ItemFlag.entries.toTypedArray())
            .build()
        val meta = wand.itemMeta
        meta.persistentDataContainer.set(itemKey, PersistentDataType.STRING, FLMS_WAND_TAG)
        wand.itemMeta = meta
        return wand
    }

    companion object {
        val FILLER = ItemStackBuilder(Material.GRAY_STAINED_GLASS_PANE)
            .name("")
            .build()
            get() {
                return field.clone()
            }

        val LEFT_POINTER = ItemStackBuilder(Material.ARROW)
            .name("${FLMS_ORANGE}<- Previous Page")
            .lore("${FLMS_YELLOW}<b>CLICK TO VIEW")
            .build()
            get() {
                return field.clone()
            }

        val RIGHT_POINTER = ItemStackBuilder(Material.ARROW)
            .name("${FLMS_ORANGE}Next Page ->")
            .lore("${FLMS_YELLOW}<b>CLICK TO VIEW")
            .build()
            get() {
                return field.clone()
            }

        val BACK_BUTTON = ItemStackBuilder(Material.BARRIER)
            .name("${FLMS_RED}Go Back")
            .lore("${FLMS_LIGHT_RED}<b>CLICK TO GO")
            .build()
            get() {
                return field.clone()
            }

        val JUMP_FIRST = ItemStackBuilder(Material.SPECTRAL_ARROW)
            .name("${FLMS_ORANGE}<b><i>Jump to First -->")
            .lore("${FLMS_YELLOW}<b>CLICK TO VIEW")
            .build()
            get() {
                return field.clone()
            }

        val JUMP_LAST = ItemStackBuilder(Material.SPECTRAL_ARROW)
            .name("${FLMS_ORANGE}<b><i><-- Jump to Last")
            .lore("${FLMS_YELLOW}<b>CLICK TO VIEW")
            .build()
            get() {
                return field.clone()
            }

        // TODO update wand icon descriptions
        val BLOCK_CONFIG_ICON = ItemStackBuilder(Material.BEDROCK)
            .name("${FLMS_ORANGE}Edit Custom Blocks")
            .lore(
                "",
                "${FLMS_LIGHT_YELLOW}Manipulate all types of blocks by",
                "${FLMS_LIGHT_YELLOW}setting their hardness values, changing",
                "${FLMS_LIGHT_YELLOW}what tools are the best at breaking them,",
                "${FLMS_LIGHT_YELLOW}and controlling what items they drop.",
                "",
                "${FLMS_YELLOW}<b>CLICK TO OPEN"
            )
            .build()
            get() {
                return field.clone()
            }

        val TOOL_CONFIG_ICON = ItemStackBuilder(Material.GOLDEN_PICKAXE)
            .name("${FLMS_ORANGE}Edit Custom Tools")
            .lore(
                "",
                "${FLMS_LIGHT_YELLOW}Create your own tools from any",
                "${FLMS_LIGHT_YELLOW}item in the game and define how",
                "${FLMS_LIGHT_YELLOW}strong they are with breaking power",
                "${FLMS_LIGHT_YELLOW}and enchantments.",
                "",
                "${FLMS_YELLOW}<b>CLICK TO OPEN"
            )
            .build()
            get() {
                return field.clone()
            }

        val EFFECT_CONFIG_ICON = ItemStackBuilder(Material.POTION)
            .name("${FLMS_ORANGE}Edit Player Effects")
            .lore(
                "",
                "${FLMS_LIGHT_YELLOW}Vanilla haste and mining fatigue",
                "${FLMS_LIGHT_YELLOW}won't work with this plugin, so you",
                "${FLMS_LIGHT_YELLOW}can modify those here instead or with",
                "${FLMS_LIGHT_YELLOW}the ${FLMS_YELLOW}/flms effect ${FLMS_LIGHT_YELLOW}command.",
                "",
                "${FLMS_YELLOW}<b>CLICK TO OPEN"
            )
            .build()
            get() {
                return field.clone()
            }

        val CREATION_ICON = ItemStackBuilder(Material.CRAFTING_TABLE)
            .name("${FLMS_GREEN}Create New...")
            .lore("${FLMS_LIGHT_GREEN}<bold>CLICK TO CREATE")
            .build()
            get() {
                return field.clone()
            }

        val FINISH_ICON = ItemStackBuilder(Material.WRITABLE_BOOK)
            .name("${FLMS_GREEN}Finish")
            .lore("${FLMS_LIGHT_GREEN}<bold>CLICK TO CREATE")
            .build()
            get() {
                return field.clone()
            }

        val POST_BREAK_EMPTY_ICON = ItemStackBuilder(Material.WHITE_STAINED_GLASS_PANE)
            .name("${FLMS_YELLOW}Post-Break Type")
            .lore(
                "${FLMS_WHITE}Drag an item here to set",
                "${FLMS_WHITE}the post-type for this block.",
                "",
                "${FLMS_GRAY}The block's post-break type is",
                "${FLMS_GRAY}the type of block it will turn into",
                "${FLMS_GRAY}after it's broken."
            )
            .build()
            get() {
                return field.clone()
            }

        val BLOCK_TYPE_EMPTY_ICON = ItemStackBuilder(Material.WHITE_STAINED_GLASS_PANE)
            .name("${FLMS_YELLOW}Block Type ${FLMS_RED}*REQUIRED*")
            .lore(
                "${FLMS_WHITE}Drag an item here to set",
                "${FLMS_WHITE}the type for this block.",
                "",
                "${FLMS_GRAY}Every custom block must have",
                "${FLMS_GRAY}a unique type. If another block",
                "${FLMS_GRAY}is using your block, you won't be",
                "${FLMS_GRAY}able to put it in here."
            )
            .build()
            get() {
                return field.clone()
            }

        val NOT_READY_ICON = ItemStackBuilder(Material.BOOK)
            .name("Not ready to finish")
            .lore(
                "",
                "${FLMS_GRAY}Make sure all the block",
                "${FLMS_GRAY}values are set correctly."
            )
            .build()
            get() {
                return field.clone()
            }

        val CANCEL_ICON = ItemStackBuilder(Material.RED_WOOL)
            .name("${FLMS_RED}Cancel")
            .build()
            get() {
                return field.clone()
            }

        val CONFIRM_ICON = ItemStackBuilder(Material.LIME_WOOL)
            .name("${FLMS_GREEN}Confirm")
            .build()
            get() {
                return field.clone()
            }

    }

    inner class Enchanter {

        fun effApply(item: ItemStack, level: Short, show: Boolean) {
            require(!item.type.isAir)

            val meta = item.itemMeta
            if (hasEff(item) && level == 0.toShort())
                meta.persistentDataContainer.remove(effKey)
            else
                meta.persistentDataContainer.set(effKey, PersistentDataType.SHORT, level)

            updateLore(meta, level, show)
            updateGlint(meta, level, show)

            item.itemMeta = meta
        }

        fun updateLore(meta: ItemMeta, level: Short, show: Boolean) {
            val cleanLore =
                (if (meta.hasLore())
                    meta.lore()
                else
                    mutableListOf<Component>())!!

            for (line in cleanLore)
                if (line.toString().contains(EFFICIENCY_PREFIX)) {
                    cleanLore.remove(line)
                    break
                }

            if (show && level != 0.toShort())
                cleanLore.addFirst(msgFormat("<gray><!i>$EFFICIENCY_PREFIX$level"))

            meta.lore(cleanLore)
        }


        @Suppress("UsePropertyAccessSyntax") // doesn't work with enchantment glint override
        fun updateGlint(meta: ItemMeta, level: Short, show: Boolean) {
            if (meta.hasEnchants())
                return
            if (level == 0.toShort() || !show)
                meta.setEnchantmentGlintOverride(null)
            else
                meta.setEnchantmentGlintOverride(true)
        }

        fun hasEff(item: ItemStack): Boolean {
            return !item.type.isAir && item.itemMeta.persistentDataContainer.has(effKey)
        }

        fun level(item: ItemStack): Short {
            if (!hasEff(item))
                return 0
            return item.itemMeta.persistentDataContainer.get(effKey, PersistentDataType.SHORT)!!
        }

    }
}


